test-fixture/src/lib.rs

```diff
@@ -70,14 +70,26 @@ pub fn default_server() -> Connection {
         .expect("create a default server")
 }
 
+/// If state is AuthenticationNeeded call authenticated(). This funstion will consume
+/// all outstanding events on the connection.
+pub fn maybe_autenticate(conn: &mut Connection) -> bool {
+    let authentication_needed = |e| matches!(e, ConnectionEvent::AuthenticationNeeded);
+    if conn.events().any(authentication_needed) {
+        conn.authenticated(0, now());
+        return true;
+    }
+    false
+}
+
 pub fn handshake(client: &mut Connection, server: &mut Connection) {
     let mut a = client;
     let mut b = server;
     let mut datagram = None;
     let is_done = |c: &Connection| matches!(c.state(), State::Connected | State::Closing { .. } | State::Closed(..));
-    while !is_done(a) {
+    while !is_done(a) || !is_done(b) {
```

Comment:
This change isn't necessary if you change the statement below to:

```rust
if !maybe_authenticate(a) {
  // If we authenticated a, keep a in place.
  mem::swap(&mut a, &mut b);
}
```

Or you could do as above: https://github.com/mozilla/neqo/pull/106/files#diff-46de255ccea3f53c2144ec0036f9698bR1963

Reply:
fix as above.